traversal/scanner/GraphIterator.java

```diff
@@ -226,27 +225,32 @@ private void traverse(ProcedureVertex<?, ?> procedureVertex) {
     }
 
     private void success(VertexTraverser vertexTraverser) {
-        for (ProcedureVertex<?, ?> vertex : vertexTraverser.procedureVertex.dependents()) {
-            toTraverse.add(vertex);
-            vertexTraversers.get(vertex).clear();
-        }
-        for (ProcedureVertex<?, ?> vertex : vertexTraverser.implicitDependents) {
-            toTraverse.add(vertex);
-            vertexTraversers.get(vertex).clear();
-        }
+        if (vertexTraverser.procedureVertex.order() + 1 == procedure.vertexCount()) return;
+        ProcedureVertex<?, ?> next = procedure.vertex(vertexTraverser.procedureVertex.order() + 1);
+        toTraverse.add(next);
+        vertexTraversers.get(next).clear();
     }
 
     private void failed(VertexTraverser vertexTraverser) {
         toRevisit.addAll(vertexTraverser.procedureVertex.dependees());
         toRevisit.addAll(vertexTraverser.implicitDependees);
+        if (!vertexTraverser.anyAnswerFound() && !vertexTraverser.procedureVertex.ins().isEmpty()) {
+            // short circuit everything not required to be re-visited
+            for (int i = vertexTraverser.lastDependee.order() + 1; i < vertexTraverser.procedureVertex.order(); i++) {
+                ProcedureVertex<?, ?> skip = procedure.vertex(i);
+                toRevisit.remove(skip);
+                vertexTraversers.get(skip).clear();
+            }
+        }
```

Comment:
when we can short-circuit backtrack to the dependees (eg. when we failed to find a single answer for the traverser), we need to clear all the intermediate vertices that we are skipping past.

Reply:
that is the correct solution, yes!

Reply:
In the previous version, we shouldn't clear `2` and `3` when `4` fails